### DISCUSSION TITLE
Does not work properly on macOS 10.13 & 10.14.

### DISCUSSION THREAD

**Discussion by hwaki** [mvp/uhubctl#418] [Q&A]
Installed libusb and uhubctl with Homebrew as instructed. The command "uhubctl" nor "uhubctl -f" does not show any hub info; not even a message such as "No compatible smart hubs detected!". The command just stuck and I have to terminate the command by ctrl-c key. Tried on two different mac (OS 10.13 and 10.14) and both act the same. The only thing that is working is the command with -h (help) and -v (version) options. What am I doing wrong?

**Comment by mvp** [comment]
Which version of uhubctl and libusb you have? You should have uhubctl 2.4.0 and libusb 1.0.26.
Unfortunately right now I only have access to MacOS 11.x and 12.x devices. I do remember that it worked fine on 10.13 (high sierra) and 10.14 (mojave) for me, but back then I used older version of libusb (1.0.21) - try linking with that.

**Reply by hwaki** [reply]
I had uhubctl 2.4.0 and libusb 1.0.26 (installed via homebrew). I uninstalled them and tried MacPorts instead and everything worked fine even with the same version of uhubctl and libusb.

**Reply by hwaki** [reply]
The command worked for a few times and then stopped working. So something else should causing this issue. What's the best way to trace the cause of the issue?

**Comment by mvp** [comment]
Perhaps you could try debugging it under gdb? Since you are running on MacOS 10.x,
you should be able to just do this:
```
git clone https://github.com/mvp/uhubctl
cd uhubctl
make
brew install gdb
gdb ./uhubctl
b main
r
<use n or s from here to debug and check variables using p>
```
This will not work on MacOS 11.x+ out of the box - you have to codesign gdb, which is quite painful.

However, you can always sprinkle `uhubctl.c` with `printf()` statements here and there and hopefully understand where it gets stuck.

**Reply by mvp** [reply]
If `gdb` gives you issues with codesigning, I checked that `lldb` works instead just fine (`lldb` is LLVM's `gdb` replacement provided by Apple and is already codesigned).
You can debug with:
```
git clone https://github.com/mvp/uhubctl
cd uhubctl
make
lldb ./uhubctl
b main
r
<use n or s from here to debug and check variables using p>
```

**Reply by hwaki** [reply]
I'm still working on the debuging but seems like the program gets stuck at following libusb function call when it tries to get hub info.  
\* By the way, tested uhubctl (via homebrew) with a few different macs running macOS Monterey/Catalina/Mojave/High Sierra, and it only worked under Monterey.

```c
len = libusb_control_transfer(devh,
            LIBUSB_ENDPOINT_IN | LIBUSB_REQUEST_TYPE_CLASS
                               | LIBUSB_RECIPIENT_DEVICE, /* hub status */
            LIBUSB_REQUEST_GET_DESCRIPTOR,
            desc_type << 8,
            0,
            buf, sizeof(buf),
            USB_CTRL_GET_TIMEOUT
        );
```
====================================
```
* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
  * frame #0: 0x00007fff7366336a libsystem_kernel.dylib`poll + 10
    frame #1: 0x00000001000f7a70 libusb-1.0.0.dylib`usbi_wait_for_events + 94
    frame #2: 0x00000001000f661b libusb-1.0.0.dylib`handle_events + 200
    frame #3: 0x00000001000f62bf libusb-1.0.0.dylib`libusb_handle_events_timeout_completed + 328
    frame #4: 0x00000001000f6af2 libusb-1.0.0.dylib`libusb_handle_events_completed + 37
    frame #5: 0x00000001000f7541 libusb-1.0.0.dylib`sync_transfer_wait_for_completion + 72
    frame #6: 0x00000001000f73f0 libusb-1.0.0.dylib`libusb_control_transfer + 284
    frame #7: 0x000000010000399f uhubctl`get_hub_info(dev=0x00000001005006f0, info=0x00007ffeefbff4a8) at uhubctl.c:391:15
    frame #8: 0x0000000100002dae uhubctl`usb_find_hubs at uhubctl.c:713:14
    frame #9: 0x00000001000019a9 uhubctl`main(argc=1, argv=0x00007ffeefbff950) at uhubctl.c:1017:10
    frame #10: 0x00007fff735263d5 libdyld.dylib`start + 1
```
====================================
```* thread #1, queue = 'com.apple.main-thread', stop reason = signal SIGSTOP
    frame #0: 0x00007fff7366336a libsystem_kernel.dylib`poll + 10
libsystem_kernel.dylib`poll:
->  0x7fff7366336a <+10>: jae    0x7fff73663374            ; <+20>
    0x7fff7366336c <+12>: movq   %rax, %rdi
    0x7fff7366336f <+15>: jmp    0x7fff7365b41d            ; cerror
    0x7fff73663374 <+20>: retq   
```

**Reply by mvp** [reply]
It seems that this is issue with your hub or libusb/MacOS inability to get hub descriptors. This function simply sends control request to the hub to identify itself - like tell vendor/product or serial number strings. Can you try with different hub(s)?

**Reply by hwaki** [reply]
I tested it without an external USB hub connected but the results were the same. I haven't been able to figure out the cause of the hang with libusb 1.0.26 but I tested uhubctl  2.4 with older version of libusb's and here are the result. 

uhubctl 2.4 worked with following libusb version 
|macOS ver | libusb ver    |
|----------|--------------|
|Monterey | libusb 1.0.26|
|Catalina | libusb 1.0.23|
|Mojave	| libusb 1.0.22|
|High Sierra | libusb 1.0.22|
